src/mds/MDBalancer.cc

```diff
@@ -337,6 +340,7 @@ void MDBalancer::queue_split(CDir *dir)
 
 void MDBalancer::queue_merge(CDir *dir)
 {
+  assert(mds->mdsmap->allows_dirfrags());
```

Comment:
This is incorrect, allows_dirfrags controls whether splitting is permitted.  If it is switched off again after some splits have happened, merges are allowed even though allow_dirfrags is false